bpf/pca.h

```diff
@@ -4,36 +4,45 @@
 #include "utils.h"
 
 static int attach_packet_payload(void *data, void *data_end, struct __sk_buff *skb) {
-    payload_meta meta;
-    u64 flags = BPF_F_CURRENT_CPU;
+    payload_meta *e;
     // Enable the flag to add packet header
     // Packet payload follows immediately after the meta struct
     u32 packetSize = (u32)(data_end - data);
 
     // Record the current time.
     u64 current_time = bpf_ktime_get_ns();
 
+    e = bpf_ringbuf_reserve(&packet_record, sizeof(payload_meta), 0);
+    if (!e) {
+        return TC_ACT_UNSPEC;
+    }
+
     // For packets which are allocated non-linearly struct __sk_buff does not necessarily
     // has all data lined up in memory but instead can be part of scatter gather lists.
     // This command pulls data from the buffer but incurs data copying penalty.
     if (packetSize <= skb->len) {
         packetSize = skb->len;
         if (bpf_skb_pull_data(skb, skb->len)) {
+            // Release reserved ringbuf location
+            bpf_ringbuf_discard(e, 0);
             return TC_ACT_UNSPEC;
         };
     }
-    // Set flag's upper 32 bits with the size of the paylaod and the bpf_perf_event_output will
-    // attach the specified amount of bytes from packet to the perf event
-    // https://github.com/xdp-project/xdp-tutorial/tree/9b25f0a039179aca1f66cba5492744d9f09662c1/tracing04-xdp-tcpdump
-    flags |= (u64)packetSize << 32;
-
-    meta.if_index = skb->ifindex;
-    meta.pkt_len = packetSize;
-    meta.timestamp = current_time;
-    if (bpf_perf_event_output(skb, &packet_record, flags, &meta, sizeof(meta))) {
-        return TC_ACT_OK;
+
+    e->if_index = skb->ifindex;
+    e->pkt_len = packetSize;
+    e->timestamp = current_time;
+    if (packetSize > MAX_PAYLOAD_SIZE) {
+        packetSize = MAX_PAYLOAD_SIZE;
     }
-    return TC_ACT_UNSPEC;
+    if (packetSize > 0 && bpf_skb_load_bytes(skb, 0, e->payload, packetSize)) {
```

Comment:
since you've called `bpf_skb_pull_data` here the skb data has been linearized already you may as well just use `memcpy`, which might save you some instructions. otherwise you could just call `bpf_skb_load_bytes` and not `bpf_skb_pull_data`.

/cc @tohojo to check my understanding is correct.

Reply:
I don't think memcpy will compile with variable length